src/envoy/tcp/sni_verifier/sni_verifier.cc

```diff
@@ -0,0 +1,184 @@
+        }
+
+        // Return an error to stop the handshake; we have what we wanted
+        // already.
+        *out_alert = SSL_AD_USER_CANCELLED;
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+      });
+}
+
+bssl::UniquePtr<SSL> Config::newSsl() {
+  return bssl::UniquePtr<SSL>{SSL_new(ssl_ctx_.get())};
+}
+
+Filter::Filter(const ConfigSharedPtr config)
+    : config_(config),
+      ssl_(config_->newSsl()),
+      buf_(std::make_unique<uint8_t[]>(config_->maxClientHelloSize())) {
+  SSL_set_accept_state(ssl_.get());
+}
+
+Network::FilterStatus Filter::onData(Buffer::Instance& data, bool) {
+  ENVOY_CONN_LOG(trace, "SniVerifier: got {} bytes",
+                 read_callbacks_->connection(), data.length());
+  if (done_) {
+    return is_match_ ? Network::FilterStatus::Continue
+                     : Network::FilterStatus::StopIteration;
+  }
+
+  size_t left_space_in_buf = config_->maxClientHelloSize() - read_;
+  size_t data_to_read =
+      (data.length() < left_space_in_buf) ? data.length() : left_space_in_buf;
+  data.copyOut(0, data_to_read, buf_.get() + read_);
+
+  auto start_handshake_data =
+      restart_handshake_ ? buf_.get() : buf_.get() + read_;
+  auto handshake_size =
+      restart_handshake_ ? read_ + data_to_read : data_to_read;
+
+  read_ += data_to_read;
+  parseClientHello(start_handshake_data, handshake_size);
+
+  return is_match_ ? Network::FilterStatus::Continue
+                   : Network::FilterStatus::StopIteration;
+}
+
+void Filter::onServername(absl::string_view servername) {
+  if (!servername.empty()) {
+    config_->stats().inner_sni_found_.inc();
+    absl::string_view outer_sni =
+        read_callbacks_->connection().requestedServerName();
+
+    is_match_ = (servername == outer_sni);
+    if (!is_match_) {
+      config_->stats().snis_do_not_match_.inc();
+    }
+    ENVOY_LOG(
+        debug,
+        "sni_verifier:onServerName(), inner SNI: {}, outer SNI: {}, match: {}",
+        servername, outer_sni, is_match_);
+  } else {
+    config_->stats().inner_sni_not_found_.inc();
+  }
+  clienthello_success_ = true;
+}
+
+void Filter::done(bool success) {
+  ENVOY_LOG(trace, "sni_verifier: done: {}", success);
+  done_ = true;
+  if (success) {
+    read_callbacks_->continueReading();
+  }
+}
+
+void Filter::parseClientHello(const void* data, size_t len) {
+  // Ownership is passed to ssl_ in SSL_set_bio()
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(data, len));
+
+  // Make the mem-BIO return that there is more data
+  // available beyond it's end
+  BIO_set_mem_eof_return(bio.get(), -1);
+
+  SSL_set_bio(ssl_.get(), bio.get(), bio.get());
+  bio.release();
+
+  restart_handshake_ = false;
+  SSL_set_app_data(ssl_.get(), this);
+  int ret = SSL_do_handshake(ssl_.get());
+
+  // reset the app data
+  SSL_set_app_data(ssl_.get(), nullptr);
+
+  // This should never succeed because an error is always returned from the SNI
+  // callback.
+  ASSERT(ret <= 0);
+  switch (SSL_get_error(ssl_.get(), ret)) {
```

Comment:
Do we need to clear SSL errors by calling ERR_clear_error()?

Reply:
@qiwzhang I have added it to be on the safe side in https://github.com/istio/proxy/pull/2050/commits/607ee02e0fe51e72d5e3327c54ef0d3765596bce.